In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [21]:
df=pd.read_csv(r"H:\datasets\ecommerce-sales.csv")

In [27]:
df.head(20)

,CID,TID,Gender,Age Group,Purchase Date,Product Category,Discount Availed,Discount Name,Discount Amount (INR),Gross Amount,Net Amount,Purchase Method,Location
0,943146,5876328741,Female,25-45,30/08/2023 20:27:08,Electronics,Yes,FESTIVE50,64.30,725.304000,661.004000,Credit Card,Ahmedabad
1,180079,1018503182,Male,25-45,23/02/2024 09:33:46,Electronics,Yes,SEASONALOFFER21,175.19,4638.991875,4463.801875,Credit Card,Bangalore
2,337580,3814082218,Other,60 and above,06/03/2022 09:09:50,Clothing,Yes,SEASONALOFFER21,211.54,1986.372575,1774.832575,Credit Card,Delhi
3,180333,1395204173,Other,60 and above,04/11/2020 04:41:57,Sports & Fitness,No,NaN,0.00,5695.612650,5695.612650,Debit Card,Delhi
4,447553,8009390577,Male,18-25,31/05/2022 17:00:32,Sports & Fitness,Yes,WELCOME5,439.92,2292.651500,1852.731500,Credit Card,Delhi
5,200614,3994452858,Male,18-25,12/07/2021 15:10:27,Clothing,Yes,FESTIVE50,127.01,3649.397400,3522.387400,Credit Card,Delhi
6,668787,9460023916,Male,25-45,08/10/2023 16:59:52,Electronics,No,NaN,0.00,5593.935600,5593.935600,PhonePe UPI,Delhi
7,933551,6877673032,Female,18-25,14/12/2022 20:13:45,Electronics,No,NaN,0.00,1901.796400,1901.796400,Credit Card,Delhi
8,414167,2841266620,Male,18-25,29/04/2020 09:10:46,Pet Care,No,NaN,0.00,3961.671000,3961.671000,Debit Card,Delhi
9,889269,1342270638,Other,60 and above,18/12/2023 19:04:02,Clothing,Yes,SAVE10,437.50,1497.350400,1059.850400,Credit Card,Bangalore


In [83]:
df.shape

(55000, 13)

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55000 entries, 0 to 54999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   CID                    55000 non-null  int64         
 1   TID                    55000 non-null  int64         
 2   Gender                 55000 non-null  object        
 3   Age Group              55000 non-null  object        
 4   Purchase Date          55000 non-null  datetime64[ns]
 5   Product Category       55000 non-null  object        
 6   Discount Availed       55000 non-null  object        
 7   Discount Name          27415 non-null  object        
 8   Discount Amount (INR)  55000 non-null  float64       
 9   Gross Amount           55000 non-null  float64       
 10  Net Amount             55000 non-null  float64       
 11  Purchase Method        55000 non-null  object        
 12  Location               55000 non-null  object        
dtypes

In [24]:
df.describe()

,CID,TID,Discount Amount (INR),Gross Amount,Net Amount
count,55000.000000,5.500000e+04,55000.000000,55000.000000,55000.000000
mean,551245.593891,5.504740e+09,136.986796,3012.936606,2875.949810
std,260603.330337,2.594534e+09,165.375502,1718.431066,1726.127778
min,100009.000000,1.000163e+09,0.000000,136.454325,-351.119775
25%,323717.000000,3.252604e+09,0.000000,1562.111325,1429.551863
50%,550088.500000,5.498383e+09,0.000000,2954.266150,2814.910875
75%,776955.750000,7.747933e+09,274.115000,4342.221675,4211.407838
max,999996.000000,9.999393e+09,500.000000,8394.825600,8394.825600


In [37]:
print(df.dtypes)

CID                               int64
TID                               int64
Gender                           object
Age Group                        object
Purchase Date            datetime64[ns]
Product Category                 object
Discount Availed                 object
Discount Name                    object
Discount Amount (INR)           float64
Gross Amount                    float64
Net Amount                      float64
Purchase Method                  object
Location                         object
dtype: object


In [36]:
df["Purchase Date"]=pd.to_datetime(df["Purchase Date"])

C:\Users\Arsalan Wani\AppData\Local\Temp\ipykernel_5728\457761688.py:1: UserWarning: Parsing dates in %d/%m/%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df["Purchase Date"]=pd.to_datetime(df["Purchase Date"])


In [48]:
print(df.isna().sum())

CID                          0
TID                          0
Gender                       0
Age Group                    0
Purchase Date                0
Product Category             0
Discount Availed             0
Discount Name            27585
Discount Amount (INR)        0
Gross Amount                 0
Net Amount                   0
Purchase Method              0
Location                     0
dtype: int64


In [50]:
df.duplicated().sum()

np.int64(0)

In [51]:
df["Discount Availed"].unique()

array(['Yes', 'No'], dtype=object)

In [31]:
df[(df["Discount Availed"]=="Yes")&(df["Discount Name"].isna())]

,CID,TID,Gender,Age Group,Purchase Date,Product Category,Discount Availed,Discount Name,Discount Amount (INR),Gross Amount,Net Amount,Purchase Method,Location


In [32]:
df[(df["Discount Availed"]=="No")&(df["Discount Name"])]

,CID,TID,Gender,Age Group,Purchase Date,Product Category,Discount Availed,Discount Name,Discount Amount (INR),Gross Amount,Net Amount,Purchase Method,Location


Discount Data Validation

Finding:
The Discount Name field is conditionally populated based on Discount Availed.
- Discount availed → discount name is present.
- No discount availed → discount name is missing.
- Both consistency checks returned 0 exceptions.

In [38]:
df["Gender"].value_counts()

Gender
Female    18454
Other     18450
Male      18096
Name: count, dtype: int64

In [40]:
df["Age Group"].value_counts()

Age Group
25-45           22010
18-25           16431
45-60           11104
under 18         2771
60 and above     2684
Name: count, dtype: int64

In [43]:
df["Age Group"].isna().sum()

np.int64(0)

In [41]:
df["Product Category"].value_counts()

Product Category
Electronics          16574
Clothing             10968
Beauty and Health     8332
Sports & Fitness      5557
Home & Kitchen        5489
Books                 2762
Other                 2171
Pet Care              1618
Toys & Games          1529
Name: count, dtype: int64

In [46]:
df["Purchase Method"].value_counts()

Purchase Method
Credit Card           22096
Debit Card            13809
Net Banking            5485
International Card     2815
Cash on Delivery       2768
PhonePe UPI            2683
Paytm UPI              2674
Google Pay UPI         2670
Name: count, dtype: int64

In [47]:
df["Location"].value_counts()

Location
Mumbai       11197
Delhi        10799
Bangalore     8249
Hyderabad     5545
Chennai       4368
Pune          3781
Ahmedabad     2785
Kolkata       2709
Jaipur        1678
Lucknow       1136
Other         1046
Varanasi       606
Dehradun       553
Srinagar       548
Name: count, dtype: int64

In [53]:
df.columns

Index(['CID', 'TID', 'Gender', 'Age Group', 'Purchase Date',
       'Product Category', 'Discount Availed', 'Discount Name',
       'Discount Amount (INR)', 'Gross Amount', 'Net Amount',
       'Purchase Method', 'Location'],
      dtype='object')

In [60]:
(df['Discount Amount (INR)']<0).sum()

np.int64(0)

In [61]:
(df["Gross Amount"]<0).sum()

np.int64(0)

In [64]:
(df["Net Amount"]<0).sum()

np.int64(613)

In [67]:
df[df["Net Amount"]<0][['Product Category', 'Discount Availed', 'Discount Name',
       'Discount Amount (INR)', 'Gross Amount', 'Net Amount',
       'Purchase Method']]

,Product Category,Discount Availed,Discount Name,Discount Amount (INR),Gross Amount,Net Amount,Purchase Method
311,Beauty and Health,Yes,NEWYEARS,469.58,279.258000,-190.322000,Debit Card
383,Electronics,Yes,NEWYEARS,358.58,204.886500,-153.693500,Debit Card
398,Home & Kitchen,Yes,WELCOME5,385.36,246.808100,-138.551900,Debit Card
489,Home & Kitchen,Yes,NEWYEARS,427.33,348.448275,-78.881725,Credit Card
564,Electronics,Yes,SAVE10,429.49,414.914325,-14.575675,Credit Card
...,...,...,...,...,...,...,...
54271,Electronics,Yes,SEASONALOFFER21,443.65,379.287000,-64.363000,Debit Card
54280,Home & Kitchen,Yes,SEASONALOFFER21,305.25,178.402875,-126.847125,Paytm UPI
54378,Books,Yes,SAVE10,212.69,185.241000,-27.449000,PhonePe UPI
54536,Electronics,Yes,FESTIVE50,492.84,216.315000,-276.525000,Google Pay UPI


In [71]:
(df[df["Net Amount"]<0]["Discount Amount (INR)"]>df[df["Net Amount"]<0]["Gross Amount"]).sum()

np.int64(613)

Data Quality Check:

613 records have a negative Net Amount. Further investigation showed that in all 613 records, the Discount Amount is greater than the Gross Amount, resulting in a negative Net Amount.

In [73]:
(df["Gross Amount"]-df["Discount Amount (INR)"]==df["Net Amount"]).sum()

np.int64(48375)

In [75]:
np.isclose(df["Gross Amount"]-df["Discount Amount (INR)"],df["Net Amount"]).sum()

np.int64(55000)

In [76]:
df[~np.isclose(df["Gross Amount"]-df["Discount Amount (INR)"],df["Net Amount"])]

,CID,TID,Gender,Age Group,Purchase Date,Product Category,Discount Availed,Discount Name,Discount Amount (INR),Gross Amount,Net Amount,Purchase Method,Location


In [77]:
(df["Discount Amount (INR)"]/df["Gross Amount"]*100).describe()

count    55000.000000
mean         8.701620
std         20.260738
min          0.000000
25%          0.000000
50%          0.000000
75%          9.289298
max        345.314905
dtype: float64

In [78]:
(df[df["Net Amount"] < 0]["Discount Amount (INR)"] /
 df[df["Net Amount"] < 0]["Gross Amount"] * 100).describe()

count    613.000000
mean     147.248697
std       41.505388
min      100.098019
25%      114.931778
50%      137.990495
75%      167.002623
max      345.314905
dtype: float64

In [79]:
df[df["Net Amount"]<0]["Discount Name"].value_counts()

Discount Name
NEWYEARS           183
SEASONALOFFER21    148
SAVE10              96
WELCOME5            95
FESTIVE50           91
Name: count, dtype: int64

Finding: 
613 transactions have negative Net Amount because their Discount Amount exceeds their Gross Amount. The Net Amount calculation itself is consistent across the dataset.



In [82]:
df[["Discount Amount (INR)","Gross Amount","Net Amount"]].describe()

,Discount Amount (INR),Gross Amount,Net Amount
count,55000.000000,55000.000000,55000.000000
mean,136.986796,3012.936606,2875.949810
std,165.375502,1718.431066,1726.127778
min,0.000000,136.454325,-351.119775
25%,0.000000,1562.111325,1429.551863
50%,0.000000,2954.266150,2814.910875
75%,274.115000,4342.221675,4211.407838
max,500.000000,8394.825600,8394.825600
